File: crates/resin-core/src/strategy.rs
//! B-class strategy type — the shell-facing strategy vocabulary.
//!
//! the vocabulary
//! CONVERGED from six display-only shell options onto Resin's three real
//! egress-selection policies. `StrategyId` now holds exactly the values
//! Resin supports — BALANCED / PREFER_LOW_LATENCY / PREFER_IDLE_IP — and the
//! former six names survive ONLY as accepted input spellings, so an existing
//! whitebox file keeps loading. That is the Expand step of the industrial
//! Expand -> Migrate -> Contract sequence (readers understand both spellings
//! before writers emit the new one); the one-time rewrite of the stored values
//! is `strategy_engine::migrate_b_class_values`, and the Contract step is the
//! removal of the six shell options from the UI and API surface.
//!
//! Ponytail: do NOT invent a parallel node-picker. Resin already owns
//! allocation via `allocation_policy`.
//!
//! The former catalog face (parse aliases, the Rust-side
//! strategy->allocation_policy mapping, the protocol_weight table,
//! strategy_catalog/StrategyInfo) had no consumer outside its own tests and
//! was deleted (ADR-0050 direction). With
//! the remaining many-to-one mapping collapsed to the IDENTITY: the
//! whitebox stores the Resin wire value verbatim, so the desired-vs-observed
//! comparison in `snapshot::merge_strategies` needs no second mapping table,
//! and the translation left in `src/lib/strategy.ts` is display-only.

use serde::{Deserialize, Serialize};

/// The legacy six-option shell catalog (pre-round-8). Accepted as INPUT only;
/// NEVER produced by serialization. Kept as a named list so the one-time
/// migration and its tests share ONE source of truth for "what counts as a
/// legacy token".
pub const LEGACY_B_CLASS_TOKENS: [&str; 6] = [
    "random",
    "sequential",
    "latency",
    "quality",
    "bandwidth",
    "protocol_weight",
];

/// User-facing strategy ids == Resin's `allocation_policy` enum (CONTEXT.md:
/// Egress IP Policy). Stable UPPER_SNAKE wire strings: the whitebox `b_class`
/// field stores the Resin value verbatim, so the snapshot compares desired
/// against observed with no translation at all.
///
/// Deserialization is TOLERANT at the read boundary (the six withdrawn shell
/// options are mapped many-to-one onto the three real policies by the SAME
/// table the webview used before the convergence —
/// `src/lib/strategy.ts::strategyToResinPolicy` — so a not-yet-migrated
/// whitebox keeps describing the policy the shell actually PATCHed), while an
/// unknown token is REJECTED rather than silently coerced: at the persisted
/// truth boundary a closed value set is what makes drift detectable.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum StrategyId {
    #[serde(rename = "BALANCED")]
    Balanced,
    #[serde(rename = "PREFER_LOW_LATENCY")]
    PreferLowLatency,
    #[serde(rename = "PREFER_IDLE_IP")]
    PreferIdleIp,
}

impl<'de> Deserialize<'de> for StrategyId {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let raw = String::deserialize(deserializer)?;
        Self::parse(&raw).ok_or_else(|| {
            serde::de::Error::custom(format!(
                "unknown b_class value {raw:?}: expected one of                  BALANCED/PREFER_LOW_LATENCY/PREFER_IDLE_IP, or one of the                  withdrawn shell options ({})",
                LEGACY_B_CLASS_TOKENS.join(", ")
            ))
        })
    }
}

impl StrategyId {
    /// Canonical wire value — the exact string Resin accepts and reports for
    /// `allocation_policy`.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Balanced => "BALANCED",
            Self::PreferLowLatency => "PREFER_LOW_LATENCY",
            Self::PreferIdleIp => "PREFER_IDLE_IP",
        }
    }

    /// The legislated many-to-one convergence table: a canonical value or one
    /// of the six withdrawn shell options -> the real policy. Case- and
    /// whitespace-insensitive (hand-edited whitebox files). `None` for
    /// anything else — an unknown token is never silently coerced.
    pub fn parse(raw: &str) -> Option<Self> {
        match raw.trim().to_ascii_uppercase().as_str() {
            "BALANCED" => Some(Self::Balanced),
            "PREFER_LOW_LATENCY" => Some(Self::PreferLowLatency),
            "PREFER_IDLE_IP" => Some(Self::PreferIdleIp),
            // Withdrawn shell catalog (pre-round-8 whitebox files).
            "RANDOM" | "BANDWIDTH" | "PROTOCOL_WEIGHT" => Some(Self::Balanced),
            "LATENCY" => Some(Self::PreferLowLatency),
            "SEQUENTIAL" | "QUALITY" => Some(Self::PreferIdleIp),
            _ => None,
        }
    }

    /// True when `raw` is one of the six withdrawn shell options, i.e. the
    /// value still needs the one-time rewrite to its canonical spelling.
    pub fn is_legacy_token(raw: &str) -> bool {
        let token = raw.trim().to_ascii_lowercase();
        LEGACY_B_CLASS_TOKENS.contains(&token.as_str())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The convergence table, asserted row by row against the historical
    /// webview mapping (`strategyToResinPolicy`) it replaces.
    #[test]
    fn convergence_table_matches_the_withdrawn_shell_mapping() {
        for legacy in ["random", "bandwidth", "protocol_weight"] {
            assert_eq!(
                StrategyId::parse(legacy),
                Some(StrategyId::Balanced),
                "{legacy}"
            );
        }
        assert_eq!(
            StrategyId::parse("latency"),
            Some(StrategyId::PreferLowLatency)
        );
        for legacy in ["sequential", "quality"] {
            assert_eq!(
                StrategyId::parse(legacy),
                Some(StrategyId::PreferIdleIp),
                "{legacy}"
            );
        }
        for id in [
            StrategyId::Balanced,
            StrategyId::PreferLowLatency,
            StrategyId::PreferIdleIp,
        ] {
            assert_eq!(StrategyId::parse(id.as_str()), Some(id), "{}", id.as_str());
        }
        // Case/whitespace tolerance for hand-edited files.
        assert_eq!(StrategyId::parse(" balanced "), Some(StrategyId::Balanced));
        assert_eq!(
            StrategyId::parse("prefer_idle_ip"),
            Some(StrategyId::PreferIdleIp)
        );
        // Unknown tokens are never coerced.
        assert_eq!(StrategyId::parse("p2c"), None);
        assert_eq!(StrategyId::parse(""), None);
    }

    #[test]
    fn canonical_values_are_the_resin_wire_spelling() {
        assert_eq!(StrategyId::Balanced.as_str(), "BALANCED");
        assert_eq!(StrategyId::PreferLowLatency.as_str(), "PREFER_LOW_LATENCY");
        assert_eq!(StrategyId::PreferIdleIp.as_str(), "PREFER_IDLE_IP");
    }

    /// Legacy tokens deserialize (Expand step) but ALWAYS serialize back as the
    /// canonical value, so ONE read+write cycle migrates the stored spelling.
    #[test]
    fn legacy_tokens_deserialize_but_serialize_canonical() {
        for (legacy, expected) in [
            ("random", "BALANCED"),
            ("bandwidth", "BALANCED"),
            ("protocol_weight", "BALANCED"),
            ("latency", "PREFER_LOW_LATENCY"),
            ("sequential", "PREFER_IDLE_IP"),
            ("quality", "PREFER_IDLE_IP"),
        ] {
            let parsed: StrategyId =
                serde_json::from_value(serde_json::Value::String(legacy.to_string())).unwrap();
            assert_eq!(parsed.as_str(), expected, "{legacy}");
            // Serialization ALWAYS emits the canonical wire spelling.
            assert_eq!(
                serde_json::to_value(parsed).unwrap(),
                serde_json::Value::String(expected.to_string())
            );
        }
    }

    /// An unknown value fails the parse with a diagnostic naming the legal set
    /// — the closed-enum discipline that keeps drift detectable.
    #[test]
    fn unknown_token_is_rejected_not_coerced() {
        let err =
            serde_json::from_value::<StrategyId>(serde_json::Value::String("p2c".to_string()))
                .unwrap_err()
                .to_string();
        assert!(err.contains("unknown b_class value"), "{err}");
        assert!(err.contains("PREFER_LOW_LATENCY"), "{err}");
    }

    #[test]
    fn is_legacy_token_flags_exactly_the_withdrawn_catalog() {
        for token in LEGACY_B_CLASS_TOKENS {
            assert!(StrategyId::is_legacy_token(token), "{token}");
        }
        for canonical in [
            "BALANCED",
            "PREFER_LOW_LATENCY",
            "PREFER_IDLE_IP",
            "balanced",
            "p2c",
            "",
        ] {
            assert!(!StrategyId::is_legacy_token(canonical), "{canonical}");
        }
    }
}

File: src/lib/strategy.ts
/**
 * the shell strategy
 * vocabulary IS Resin's `allocation_policy` enum.
 *
 * The GUI used to expose six shell options that collapsed many-to-one onto
 * Resin's three real policies. That catalogue is withdrawn: the selector now
 * offers exactly the three values Resin supports, and the whitebox stores the
 * Resin wire value verbatim — so the mapping that used to live here
 * collapsed to the identity and there is nothing left to translate.
 *
 * The former six names are still ACCEPTED on read (`normalizeStrategy`) so a
 * whitebox file written before the convergence keeps rendering, but they are
 * never produced: the one-time rewrite to the canonical spelling lives in Rust
 * (`resin_core::strategy_engine::migrate_b_class_values`), and the Rust
 * `StrategyId` rejects an unknown token outright.
 *
 * Sole vocabulary owner: this file is the ONLY strategy <->
 * allocation_policy mapping left in the repo — used for view display labels
 * and the webview-side translation before the platform PATCH. The Rust side
 * holds the same three values as the storage type; do not reintroduce a second
 * mapping copy in Rust (ADR-0052 keeps the vocabularies separate and this file
 * is the display layer).
 */
export const STRATEGY_IDS = [
  "BALANCED",
  "PREFER_LOW_LATENCY",
  "PREFER_IDLE_IP",
] as const;

export type StrategyId = (typeof STRATEGY_IDS)[number];

/// Resin allocation_policy enum (the backend wire format) — identical to the
/// shell vocabulary after the convergence.
export type AllocationPolicy = StrategyId;

export const ALLOCATION_POLICIES: AllocationPolicy[] = [...STRATEGY_IDS];

/// The withdrawn six-option shell catalogue, accepted on read only.
const LEGACY_B_CLASS_TOKENS = [
  "random",
  "sequential",
  "latency",
  "quality",
  "bandwidth",
  "protocol_weight",
] as const;

/**
 * The legislated many-to-one convergence table: a canonical value
 * or one of the six withdrawn shell options -> the real policy. Mirrors
 * `resin_core::strategy::StrategyId::parse` row for row. Returns null for an
 * unrecognized token — never silently coerced.
 */
export function normalizeStrategy(s: string): AllocationPolicy | null {
  switch (s.trim().toUpperCase()) {
    case "BALANCED":
    case "RANDOM":
    case "BANDWIDTH":
    case "PROTOCOL_WEIGHT":
      return "BALANCED";
    case "PREFER_LOW_LATENCY":
    case "LATENCY":
      return "PREFER_LOW_LATENCY";
    case "PREFER_IDLE_IP":
    case "SEQUENTIAL":
    case "QUALITY":
      return "PREFER_IDLE_IP";
    default:
      return null;
  }
}

/// Map a strategy value -> i18n key (uses committed strategy.* keys). Legacy
/// tokens resolve to the canonical label: a withdrawn option no longer has a
/// label of its own.
export function strategyToI18nKey(s: string): string {
  switch (normalizeStrategy(s)) {
    case "BALANCED":
      return "strategy.balanced";
    case "PREFER_LOW_LATENCY":
      return "strategy.preferLowLatency";
    case "PREFER_IDLE_IP":
      return "strategy.preferIdleIp";
    default:
      return s;
  }
}

/// Map a shell strategy value -> Resin allocation_policy for the backend PATCH.
/// Identity after the convergence; kept as the single translation point AND as
/// the tolerant normalizer for a legacy value still in flight. An unknown value
/// falls back to BALANCED (Resin's own default).
export function strategyToResinPolicy(s: StrategyId | string): AllocationPolicy {
  return normalizeStrategy(s) ?? "BALANCED";
}

/// Map a Resin allocation_policy value back to the shell vocabulary for UI
/// display. Identity after the convergence.
export function mapResinToShell(p: string): StrategyId {
  return normalizeStrategy(p) ?? "BALANCED";
}

/// Validate that a string is a valid shell StrategyId (canonical values only).
export function isValidStrategyId(s: string): s is StrategyId {
  return (STRATEGY_IDS as readonly string[]).includes(s);
}

/// True when the value is one of the six withdrawn shell options (i.e. it still
/// needs the one-time rewrite to its canonical spelling).
export function isLegacyStrategyToken(s: string): boolean {
  return (LEGACY_B_CLASS_TOKENS as readonly string[]).includes(
    s.trim().toLowerCase(),
  );
}

/// the B-class badge label. The former
/// per-strategy parameter interpolation (round-robin N, latency threshold ms,
/// quality score, bandwidth weight) went away with `BClassParams` — those
/// were display-only values no backend ever read. The badge now states the ONE
/// thing that is real: which egress-selection policy is in effect.
export function bClassLabel(
  strategy: string,
  t: (key: string, opts?: Record<string, unknown>) => string,
): string {
  const normalized = normalizeStrategy(strategy);
  return normalized ? t(strategyToI18nKey(normalized)) : strategy;
}
